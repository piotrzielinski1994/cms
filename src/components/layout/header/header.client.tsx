'use client';

import { useHeaderTheme } from '@/_old/providers/HeaderTheme';
import { cn } from '@/_old/utilities/ui';
import Container from '@/components/layout/container/container';
import Section from '@/components/layout/section/section';
import LogoSvg from '@/icons/logo.svg';
import { Link } from '@/payload/locale/routing';
import { usePathname } from 'next/navigation';
import React, { useEffect, useState } from 'react';
import { HeaderClientProps } from './header.types';
import FontScaler from './scaffold/font-scaler';
import LocaleSwitcher from './scaffold/locale-switcher';
import HeaderNav from './scaffold/navbar';
import ThemeSwitcher from './scaffold/theme-switcher';

export const HeaderClient: React.FC<HeaderClientProps> = ({ data }) => {
  /* Storing the value in a useState to avoid hydration errors */
  const [theme, setTheme] = useState<string | null>(null);
  const { headerTheme, setHeaderTheme } = useHeaderTheme();
  const pathname = usePathname();

  useEffect(() => setHeaderTheme(null), [pathname, setHeaderTheme]);

  useEffect(() => {
    if (headerTheme && headerTheme !== theme) setTheme(headerTheme);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [headerTheme]);

  return (
    <Section as="header" className="py-5 bg-background1 sticky top-0 z-header">
      <Container className={cn('flex items-center')}>
        <Link href="/">
          <LogoSvg />
        </Link>
        <HeaderNav data={data} />
        <FontScaler />
        <ThemeSwitcher />
        <LocaleSwitcher />
      </Container>
    </Section>
  );
};
