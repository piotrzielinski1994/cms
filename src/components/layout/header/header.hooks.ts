import { useEffect, useState } from 'react';

const useHeaderHeight = () => {
  const [headerHeight, setHeaderHeight] = useState(0);

  useEffect(() => {
    const handleResize = () => {
      const header = document.querySelector('[data-header]') as HTMLElement;
      if (!header) return;
      setHeaderHeight(header.offsetHeight);
    };

    handleResize();
    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return headerHeight;
};

export { useHeaderHeight };
