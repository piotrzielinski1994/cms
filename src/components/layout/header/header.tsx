import { Container } from '@/components/basic/container';
import { Logo } from '@/components/basic/logo';
import { Section } from '@/components/basic/section';
import { getCachedGlobal } from '@/payload/utils/globals';
import { cn } from '@/utils/tailwind';
import { TypedLocale } from 'payload';
import FontScaler from './scaffold/font-scaler';
import LocaleSwitcher from './scaffold/locale-switcher';
import HeaderNav from './scaffold/navbar';
import ThemeSwitcher from './scaffold/theme-switcher';

type HeaderProps = {
  locale: TypedLocale;
};

const Header = async ({ locale }: HeaderProps) => {
  const header = await getCachedGlobal('header', locale)();
  return (
    <>
      <Section as="div" className="bg-black">
        <Container className={cn('flex justify-end items-center')}>
          <FontScaler />
          <ThemeSwitcher />
          <LocaleSwitcher />
        </Container>
      </Section>
      <Section as="header" className="py-5 bg-background1 sticky top-0 z-header">
        <Container className={cn('flex items-center')}>
          <Logo />
          <HeaderNav data={header} />
        </Container>
      </Section>
    </>
  );
};

export { Header };
