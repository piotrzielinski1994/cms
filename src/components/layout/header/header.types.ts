import { DataFromGlobalSlug, TypedLocale } from 'payload';

export interface HeaderProps {
  locale: TypedLocale;
}

export interface HeaderClientProps {
  data: DataFromGlobalSlug<'header'>;
}
