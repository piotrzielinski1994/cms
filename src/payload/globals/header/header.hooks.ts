import { GlobalRevalidationTag } from '@/payload/utils/globals';
import { Locale } from 'next-intl';
import { revalidateTag } from 'next/cache';
import type { GlobalAfterChangeHook } from 'payload';

const revalidateHeader: GlobalAfterChangeHook = ({ doc, req: { payload, context, locale } }) => {
  if (!context.disableRevalidate) {
    payload.logger.info(`Revalidating header`);
    const typedLocale = locale as Locale;
    const tag: GlobalRevalidationTag = `global__${typedLocale}__header`;
    revalidateTag(tag);
  }

  return doc;
};

export { revalidateHeader };
