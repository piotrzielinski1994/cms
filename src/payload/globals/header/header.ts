import { AdminTranslations, translations } from '@/config/locales.config';
import { link } from '@/payload/fields/link';
import { rebuildTag } from '@/utils/next';
import { isCollectionLocale } from '@/utils/payload';
import type { GlobalConfig } from 'payload';

const header = {
  slug: 'header',
  access: {
    read: () => true,
  },
  admin: {
    group: {
      en: translations.en.common.component.plural,
      pl: translations.pl.common.component.plural,
    },
  },
  label: ({ t }: { t: AdminTranslations }) => t('components:header'),
  fields: [
    {
      name: 'navItems',
      type: 'array',
      label: ({ t }: { t: AdminTranslations }) => t('fields:menu'),
      fields: [
        link({
          appearances: false,
          depth: 1,
        }),
      ],
      maxRows: 6,
      admin: {
        initCollapsed: true,
        components: {
          RowLabel: '@/payload/components/row-label#RowLabel',
        },
      },
    },
  ],
  hooks: {
    afterChange: [
      ({ doc, req: { payload, context, locale } }) => {
        if (!context.disableRevalidate && isCollectionLocale(locale)) {
          payload.logger.info(`Revalidating header`);
          rebuildTag(`global__${locale}__header`);
        }

        return doc;
      },
    ],
  },
} satisfies GlobalConfig;

export { header };
