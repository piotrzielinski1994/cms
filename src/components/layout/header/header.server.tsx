import { Page } from '@/payload/payload.types';
import { getCachedGlobal } from '@/payload/utils/globals';
import { Locale } from 'next-intl';
import { Header as HeaderComponent } from './header';

type HeaderProps = {
  locale: Locale;
};

const Header = async ({ locale }: HeaderProps) => {
  const header = await getCachedGlobal('header', locale)();
  const navItems = (header.navItems ?? []).map((it) => ({
    id: it.id!,
    label: it.link.label,
    path: String((it.link.reference?.value as Page).path),
  }));

  return <HeaderComponent items={navItems} />;
};

export { Header };
