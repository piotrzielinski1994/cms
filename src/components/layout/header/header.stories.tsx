import type { Meta, StoryObj } from '@storybook/react';
import { type ComponentProps } from 'react';
import { Header as HeaderComponent } from './header';

type Args = ComponentProps<typeof HeaderComponent> & {
  label?: string;
};

const meta: Meta<Args> = {
  component: HeaderComponent,
  title: 'Components/Layout/Header',
  argTypes: {
    items: { control: 'object' },
  },
  args: {
    items: [
      { id: 'nav-item-1', label: 'Link 1', path: '/' },
      { id: 'nav-item-2', label: 'Link 2', path: '/' },
      { id: 'nav-item-3', label: 'Link 3', path: '/' },
    ],
  },
  parameters: {
    nextjs: {
      appDirectory: true,
    },
  },
};

const Render = (args: Args) => {
  return <HeaderComponent {...args} />;
};

const Header: StoryObj<typeof HeaderComponent> = { render: Render };

export { Header };
export default meta;
