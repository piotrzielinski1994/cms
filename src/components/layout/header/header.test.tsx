import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { Header } from './header';

it('should have no accessibility violations', async () => {
  const { container } = render(<Header locale="en" />);
  const results = await axe(container);
  expect(results).toHaveNoViolations();
});
