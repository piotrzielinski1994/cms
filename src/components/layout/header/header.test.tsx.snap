// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should match the snapshot 1`] = `
<div>
  <div
    class="px-4 sm:px-6 grid cms-section py-2 bg-black"
  >
    <div
      class="w-full max-w-screen-2xl mx-auto cms-container flex justify-end items-center"
    >
      <button
        aria-autocomplete="none"
        aria-controls="radix-«rc»"
        aria-expanded="false"
        aria-label="Font scale switcher"
        class="p-2 text-sm text-white"
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          [Svg @/icons/font-scaler.svg]
        </span>
      </button>
      <button
        aria-autocomplete="none"
        aria-controls="radix-«rd»"
        aria-expanded="false"
        aria-label="Theme switcher"
        class="py-1.5 px-2 text-sm text-white"
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          [Svg @/icons/contrast.svg]
        </span>
      </button>
      <button
        aria-autocomplete="none"
        aria-controls="radix-«re»"
        aria-expanded="false"
        aria-label="Locale switcher"
        class="px-2 py-1 -mr-2 text-sm text-white"
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          EN
        </span>
      </button>
    </div>
  </div>
  <header
    class="px-4 sm:px-6 grid cms-section sticky top-0 z-header py-5 bg-background1 shadow-sm"
    data-header="true"
  >
    <div
      class="w-full max-w-screen-2xl mx-auto cms-container flex items-center"
    >
      <a
        aria-label="Logo"
        class="text-2xl"
        href="/"
      >
        [Svg @/icons/logo.svg]
      </a>
      <nav
        class="flex-grow flex justify-end"
      >
        <ul
          class="contents"
        >
          <li
            class="contents"
          >
            <a
              class="p-2"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="contents"
          >
            <a
              class="p-2"
              href="/about"
            >
              About
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </header>
</div>
`;
